plugins/Phaser/PhaserEffect.cpp

```diff
@@ -0,0 +1,613 @@
+		for (int i = 0; i < 2; i++)
+		{
+			if (sAbs[i] > m_currentPeak[i])
+			{
+				m_currentPeak[i] = qMin(m_currentPeak[i] * m_attCoeff, sAbs[i]);
+			}
+			else
+			if (sAbs[i] < m_currentPeak[i])
+			{
+				m_currentPeak[i] = qMax(m_currentPeak[i] * m_relCoeff, sAbs[i]);
+			}
+
+			m_currentPeak[i] = qBound(PHA_NOISE_FLOOR, m_currentPeak[i], 10.0f);
+		}
+
+		// Calculate real allpass filter frequencies
+		float lfoResults[2] = {0, 0};
+		if (enableLFO)
+		{
+			m_lfo->tick(&lfoResults[0], &lfoResults[1]);
+		}
+
+		const float realLfo[2] = {lfoResults[0] * amount, lfoResults[1] * amount};
+		const float realInFollow[2] = {m_currentPeak[0] * inFollow, m_currentPeak[1] * inFollow};
+		for (int i = 0; i < 2; i++)
+		{
+			m_displayCutoff[i] = qBound(20.f, detuneWithOctaves(cutoff, realLfo[i] + realInFollow[i]), 20000.f);
+			m_realCutoff[i] = qBound(20.f, detuneWithOctaves(cutoff, (realLfo[i] + realInFollow[i]) * cutoffControl), 20000.f);
+		}
+		
+		/*
+		Precalculate the allpass filter coefficients.
+		Because of very aggressive CPU optimization, some of
+		the coefficient names are inaccurate concerning what
+		they actually do.
+		*/
+		float b0[2];
+		float b1[2];
+		for (int b = 0; b < 2; ++b)
+		{
+			const float w0 = m_twoPiOverSr * m_realCutoff[b];
+			const float a0 = 1 + (sin(w0) / (resonance * 2.f));
+			b0[b] = (1 - (a0 - 1)) / a0;
+			b1[b] = (-2*cos(w0)) / a0;
+		}
+
+		float firstAdd[2] = {0, 0};
+		float secondAdd[2] = {0, 0};
+		float delayOut[2] = {0, 0};
+
+		for (int i = 0; i < 2; ++i)
+		{
+			// Calculate delay amount to find read location
+			float readLoc = m_filtFeedbackLoc -
+				qMin(abs(delay + ((realLfo[i] + realInFollow[i]) * delayControl)), 20.f) *
+				0.001f * Engine::mixer()->processingSampleRate();
+
+			if (readLoc < 0) {readLoc += m_delayBufSize;}
+			float readLocFrac = fraction(readLoc);
+
+			// Read value from delay buffer
+			if (readLoc < m_delayBufSize - 1)
+			{
+				delayOut[i] = m_filtDelayBuf[i][floor(readLoc)] * (1 - readLocFrac) + m_filtDelayBuf[i][ceil(readLoc)] * readLocFrac;
+			}
+			else// For when the interpolation wraps around to the beginning of the buffer
+			{
+				delayOut[i] = m_filtDelayBuf[i][m_delayBufSize - 1] * (1 - readLocFrac) + m_filtDelayBuf[i][0] * readLocFrac;
+			}
+		}
+
+		float tempAdd[2] = {0, 0};
+		switch (mode)
+		{
+			case 0: case 1:
+			{
+				tempAdd[0] = delayOut[0] * feedback;
+				tempAdd[1] = delayOut[1] * feedback;
+				break;
+			}
+			case 2:
+			{
+				tempAdd[0] = lastSecondAdd[0] * feedback;
+				tempAdd[1] = lastSecondAdd[1] * feedback;
+				break;
+			}
+			case 3:
+			{
+				tempAdd[0] = delayOut[0] * feedback;
+				tempAdd[1] = delayOut[1] * feedback;
+				for (int a = 0; a < order; ++a)
+				{
+					for (int b = 0; b < 2; ++b)
+					{
+						for (int c = 0; c < 1 + doubleVal; ++c)// Runs twice if doubleVal is true
+						{
+							tempAdd[b] = calcAllpassFilter(tempAdd[b], sample_rate, a, b, b0[b], b1[b]);
+						}
```

Comment:
Will this work as expected? Calling `calcAllpassFilter` updates the state variables in `m_filtX[a][b]` and `m_filtY[a][b]`. If you're going to run the filter twice, shouldn't you use a separate set of state variables to prevent feeding multiple samples for a filter state when only one real input sample is given?

Reply:
Also, I think the way you store state variables can be improved. Here's an example:
```cpp
struct FilterState
{
	float x[2];
	float y[2];
};
FilterState m_filt[32][2];
```
Also, I suggest you reading existing filter classes in `include/BasicFilters.h` and see if any of them fits your needs.

Reply:
> Will this work as expected?

Refer to the large comment I made explaining the various feedback circuits.  Both filters using the same buffer is intentional, for the appropriately-named "Inane" mode.  The extremely odd behavior is exactly what I was going for, and can result in some very cool resonant screeches and unexpected timbres.

Reply:
Actually, I may have misunderstood what you were talking about.
A separate set of state variables is not required, since the input `tempAdd[b]` is also the output `tempAdd[b]`.

Reply:
(Also, that all being said, the BasicFilters.h filters can't be used, or else the fan-favorite Inane mode will have to be removed due to its odd behavior, and keeping things ludicrously optimized is fun anyway.)

Reply:
...wait, I'm wrong.  You're right.  It'll need its own buffer for it to be two separate filters.  I was wondering why the results were more interesting and cool than I expected.
I'll update the tooltip so it's accurate.

Reply:
I've decided to leave the Double feature as it is.  From what I can tell, it ends up being equivalent to running 88200 Hz audio through a filter calculated for 44100 Hz audio, and then outputting 44100 Hz audio... but what matters is that it sounds cool.  :D 
My original idea also was not a standard or orthodox phaser feature (like most of the new things I added...), and once I implemented it the way I originally intended, it sounded too close to just doubling the filter order for it to be interesting at all.  
Instead of running the audio through two filters per loop, it runs every input through the filter twice.  "Double" still makes sense here, so I've decided to keep it.